docs/test_framework_integration/manual_lifecycle_control.md

```diff
@@ -48,3 +48,79 @@ The singleton container is started only once when the base class is loaded.
+
+Please keep in mind that putting the container in a static block will influence the configuration of the container.
+You will not be able to use methods such as `forResponsePredicate` by just providing a Lambda expression, you will have to use
+anonymous classes, or alternatively provide them from non-abstract class.
+This is not due to the limitation of `TestContainers`, but because of how they work under the hood, namely that lambdas get compiled
+to static methods on a class level. Since your container is in a static block, the container gets created
+before your parent and children classes get initialized and as such you cannot pass the reference to them.
+
+Therefore, once again - it is advised to use anonymous classes in such case or full predicates coming in from different, non-abstract class.
+
+Instead of:
+
+```java
+abstract class AbstractContainerBaseTest {
+
+    static final GenericContainer GENERIC_CONTAINER;
+
+    static {
+        GENERIC_CONTAINER = new GENERIC_CONTAINER(
+                new ImageFromDockerfile().waitingFor(Wait.forHttp('/path'))
+                                         .forStatusCode(200)
+                                         .forResponsePredicate(yourLambda -> yourLambda(here)) //This is never going to get executed due to NoClassDefFoundError
+        );
+        GENERIC_CONTAINER.start();
+    }
+}
+```
+
+You can do an anonymous class:
+
+```java
+abstract class AbstractContainerBaseTest {
+
+    static final GenericContainer GENERIC_CONTAINER;
+
+    static {
+        GENERIC_CONTAINER = new GENERIC_CONTAINER(
+                new ImageFromDockerfile().waitingFor(Wait.forHttp('/path'))
+                                         .forStatusCode(200)
+                                         .forResponsePredicate(new Predicate<String>() {
+                                             
+                                             @Override
+                                             public boolean test(String s) {
+                                                 return yourConditionHere;
+                                             }
+                                         }) 
+        );
+        
+        GENERIC_CONTAINER.start();
+    }
+}
+```
+Or full predicate coming in from different class:
+
+```java
+abstract class AbstractContainerBaseTest {
+
+    static final GenericContainer GENERIC_CONTAINER;
+
+    static {
+        GENERIC_CONTAINER = new GENERIC_CONTAINER(
+                new ImageFromDockerfile().waitingFor(Wait.forHttp('/path'))
+                                         .forStatusCode(200)
+                                         .forResponsePredicate(PredicateHolder.getPredicate())
+        );
+        GENERIC_CONTAINER.start();
+    }
+}
+
+public class PredicateHolder {
+    
+    public static Predicate<String> getPredicate() {
+        return yourLambda -> yourLambda(here);
+    } 
+}
+```
```

Comment:
There is an [issue](https://bugs.openjdk.org/browse/JDK-8281371) reported in JDK which describe what's happening here. So, static methods from other classes can not be called from a lambda in a static block. But, if those are used as a method reference then it will work.

Also, if a container class is created then this issue can be avoided we would encourage to do that

```java
class HWContainer extends GenericContainer<HWContainer> {
    public HWContainer() {
        super("testcontainers/helloworld");
        withExposedPorts(8080);
        waitingFor(Wait.forHttp("/ping").forResponsePredicate(x -> Util.isContains(x)));
    }
}
```

I would like to make a note about this and give the proper context and recommendations. Also, Can you please move the inline code to actual code and use codeblock in the file to display them? I think code examples would be avoided if the note is quite clear.

Note: The project is Testcontainers not TestContainers :)

Reply:
Hey! Thanks for the response.

I do firmly believe that this is not about deadlocking per se, but rather that static block gets initialized before the initialization of a class - and since the lambda gets compiled to a static field, trying to reference something that does not exist is impossible. 

Could you point me in the direction why do you believe that this is the deadlock situation?

Also, when you respond I will: 

1.  Move out the examples to examples/ module. Question here, does it need to be completely valid code with imports etc. or can I just copy this somewhat pseudocode?
2. Reference the JDK bug/discussion.
3. Would you rather me deleting some examples rather than including them at all? From your message I got a feeling you'd delete some of them.